r/R/record-batch.R

```diff
@@ -189,3 +189,52 @@ record_batch <- RecordBatch$create
 
 #' @export
 names.RecordBatch <- function(x) x$names()
+
+#' @export
+rbind.RecordBatch <- function(...) {
+  abort("Use Table$create to combine record batches")
+}
+
+cbind_check_length <- function(target_length, length, idx) {
+  if (length != target_length) {
+    abort(
+      sprintf(
+        "Non-scalar inputs must have an equal number of rows. ..1 has %i, ..%i has %i",
+        target_length,
+        idx,
+        length
+      )
+    )
+  }
+}
+
+#' @export
+cbind.RecordBatch <- function(...) {
+  inputs <- list(...)
+  num_rows <- inputs[[1]]$num_rows
+
+  batches <- imap(inputs, function(input, idx) {
+    if (inherits(input, "RecordBatch")) {
+      cbind_check_length(num_rows, input$num_rows, idx)
+      input
+    } else if (is.atomic(input) && length(input) == 1) {
+      RecordBatch$create("{idx}" := rep(input, num_rows))
+    } else {
+      tryCatch(
+        {
+          cbind_check_length(num_rows, length(input), idx)
+          RecordBatch$create("{idx}" := input)
+        },
+        error = function(err) {
+          abort(sprintf("Input ..%i cannot be converted to an Arrow Array: %s", idx, err))
+        }
+      )
+    }
```

Comment:
I think so, though need to double check that there aren't S3 dispatch surprises. I think at least in 3.6 and earlier it always dispatches to an internal implementation of cbind if one of the arguments is a data.frame.

Reply:
I got it to work on R >= 4.0. As I suspected, R 3.6 will hijack and calls to cbind that have a data.frame as an argument. I don't think there's anything we can do to prevent that. If users want to work around that, they could call `cbind.RecordBatch` directly and that will work even in old R versions.